peps/pep-0751.rst

```diff
@@ -0,0 +1,901 @@
+  locking, e.g. supporting a JSON file format which specifies wheel tags and
+  marker values much like in ``[[file-lock]]`` for which multiple files can be
+  specified, which could then be directly recorded in the corresponding
+  ``[[file-lock]]`` table (if it allowed for unambiguous per-file locking
+  environment selection)
+
+.. code-block:: JSON
+
+    {
+        "marker-values": {"<marker>": "<value>"},
+        "wheel-tags": ["<tag>"]
+    }
+
+
+Expectations for Installers
+---------------------------
+
+- Installers MAY support installation of non-binary files
+  (i.e. source distributions, source trees, and VCS), but are not required to
+- Installers MUST provide a way to avoid non-binary file installation for
+  reproducibility and security purposes
+- Installers SHOULD make it opt-in for using non-binary file installation to
+  facilitate a secure-by-default approach
+- Under per-file locking, if what to install is ambiguous then the installer
+  MUST raise an error
+
+
+Installing for per-file locking
+'''''''''''''''''''''''''''''''
+
+An example workflow is:
+
+- Iterate through each ``[[file-lock]]`` table to find the one that applies to
+  the environment being installed for
+- If no compatible environment is found an error MUST be raised
+- If multiple environments are found to be compatible then an error MUST be raised
+- For the compatible environment, iterate through each entry in ``[[package]]``
+- For each ``[[package]]`` entry, iterate through ``[[package.files]]`` to look
+- for any files with ``file-lock.name`` listed in ``package.files.lock``
+- If a file is found with a matching lock name, add it to the list of candidate
+  files to insntall and move on to the next ``[[package]]`` entry
+- If no file is found then check if ``package.vcs.lock`` contains a match (no
+  match is also acceptable)
+- If a ``[[package.files]]`` contains multiple matching entries an error MUST
+  be raised due to ambiguity for what is to be installed
+- If multiple ``[[package]]`` entries for the same package have matching files
+  an error MUST be raised due to ambiguity for what is to be installed
+- Find and verify the candidate files and/or CVS entries based on their hash or
+  commit ID as appropriate
+- If a source distribution or VCS was selected and
+  ``[[package.build-requires]]`` exists, then repeat the above process as
+  appropriate to install the build dependencies necessary to build the package
+- Install the candidate files
+
+
+Installing for package locking
+''''''''''''''''''''''''''''''
+
+An example workflow is:
+
+- Verify that the environment is compatible with
+  ``package-lock.requires-python``; if it isn't an error MUST be raised
+- Iterate through each entry in ``[package]]``
+- For each entry, if there's a ``package.marker`` key, evaluate the expression
+  - If the expression is false, then move on
+  - Otherwise the package entry must be installed somehow
```

Comment:
This needs some adjustment to render properly:
<img width="789" alt="image" src="https://github.com/user-attachments/assets/3ba051c3-d361-4e2e-99f0-74e7cbec0439">

Reply:
Lists in reST need to start and end with a blank line:

```suggestion
- For each entry, if there's a ``package.marker`` key, evaluate the expression

  - If the expression is false, then move on
  - Otherwise the package entry must be installed somehow
```